Add syntax of codecs parameter

nigelmegitt opened the pull request:
Also separate out the encouragement to support from the syntax definition and explanation as per discussion at https://github.com/w3c/tt-profile-registry/issues/71#issuecomment-682033293, and add a note about MP4's `codecs` parameter.

Resolves #71.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/tt-profile-registry/pull/78.html" title="Last updated on Sep 4, 2020, 9:54 AM UTC (a88760d)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/tt-profile-registry/78/0f9981c...a88760d.html" title="Last updated on Sep 4, 2020, 9:54 AM UTC (a88760d)">Diff</a>

nigelmegitt committed:
Add syntax of codecs parameter

Also separate out the encouragement to support from the syntax definition and explanation as per discussion at #71.

nigelmegitt committed:
Add note about MP4 codecs parameter

mikedo commented:
1. It would be good to add to the // comment about "element" that "period" is forbidden (analogous to the other comments).
2. The new note about "the other" codecs parameter should better say: "Another  codecs  parameter is defined by [ RFC 6381 ] for application/mp4 (ISO BMFF) and extended for TTML by [  iso14496-30  ]."
3. We should provide a concrete example (and not just the abstract ones).  I offered one in my earlier comment.
4. The formatting of the last line of the ABNF is broken into several lines.

cconcolato commented on index.html:
```diff
@@ -141,7 +158,34 @@ <h2>Purpose</h2>
+                                     // between each token
+
+codecs-parameter-value
+: and_group ["|" and-group]*         // no linear white-space is implied or permitted
```
`and_group` should be `and-group`

nigelmegitt replied:
Fixed in 6a651c9

nigelmegitt commented:
>  1. It would be good to add to the // comment about "element" that "period" is forbidden (analogous to the other comments).

I just traced this through, and I don't think "period" is forbidden.

The existing text in the TTML Profiles Registry says:

> An identifier must conform with the `element` non-terminal of [RFC6381] and, furthermore, may not contain any of the characters in the regular expression character class `[+|]`.

[RFC6381](https://tools.ietf.org/html/rfc6381#section-3.2) defines `element` as:

```
TOKEN is defined in [RFC2045]
...
element     := 1*octet-sim
octet-sim   := <any TOKEN character>
```

[[RFC2045]](https://tools.ietf.org/html/rfc2045) defines:

```
     token := 1*<any (US-ASCII) CHAR except SPACE, CTLs,
                 or tspecials>

     tspecials :=  "(" / ")" / "<" / ">" / "@" /
                   "," / ";" / ":" / "\" / <">
                   "/" / "[" / "]" / "?" / "="
                   ; Must be in quoted-string,
                   ; to use within parameter values

...

Note that the definition of "tspecials" is the same as the RFC 822
   definition of "specials" with the addition of the three characters
   "/", "?", and "=", and the removal of ".".
```

By my reading, this means that `token`, and therefore `element`, is permitted to contain a period `"."`. @mikedo I guess this could be the subject of a separate issue, but for the time being I won't be adding a comment about period being forbidden. Happy to change my view if I've missed something here, of course.

nigelmegitt committed:
Address review comments

nigelmegitt commented:
@mikedo @cconcolato I've addressed your review comments; I was left a bit confused by our discussion about the concrete example, since it also appears to be part of the proposal in #76; on balance I think it better addressed by a separate pull request there if that's okay.

The formatting problem appears to be an issue only with the HTML diff output, so no change needed there.

mikedo commented:
>> It would be good to add to the // comment about "element" that "period" is forbidden (analogous to the other comments).

> I just traced this through, and I don't think "period" is forbidden.

Period is a core delimiter in the 6381 codecs syntax.  Element cannot contain a period.  See 6381, section 3.4.

nigelmegitt commented:
> Period is a core delimiter in the 6381 codecs syntax. Element cannot contain a period. See 6381, section 3.4.

That section does not define `element` though, it merely uses it. Section 3.2 defines `element` as noted above. I don't know if there was a different intent in 6381 that somehow didn't make it through, but the exclusion of `"."` from `tspecials` is called out explicitly in a note, so it looks like a very deliberate choice to permit it in `element`.

Oh hang on, I missed a crucial line two below the definition of `octet-sim`:

```
      element     := 1*octet-sim
      octet-sim   := <any TOKEN character>

                  ; Within a 'codecs' parameter value, "." is reserved
                  ; as a hierarchy delimiter
```

Since we do _not_ want to allow hierarchy delimiters in the midst of identifiers, I agree we need to clarify that the intent for _our_ `codecs` parameter is the same, i.e. prohibit `"."`.

mikedo commented:
> I was left a bit confused by our discussion about the concrete example, since it also appears to be part of the proposal in #76; on balance I think it better addressed by a separate pull request there if that's okay.

I don't understand the concern, but to remove all doubt about what I propose, please add the following real life examples somewhere:

For example, the codecs string for TTML documents that conform to both IMSC1.0.1 text profile and EBU-TT-D-1 (the intersection) would be "im1t|etd1". And for TTML documents that require both IMSC1.0.1 and EBU-TT-D-1 (the union) would be "im1t+etd1".

nigelmegitt committed:
Prohibit `"."` from `element`

nigelmegitt committed:
Add concrete example note

nigelmegitt commented:
OK @mikedo please could you have a look now, I think both remaining comments are addressed.
